### DISCUSSION TITLE
help !! custom bankable not work

### DISCUSSION THREAD

**Discussion by fareign** [tttapa/Control-Surface#1072] [General]
```cpp
#include "Control_Surface.h"
BEGIN_CS_NAMESPACE

class PanJoystick : public MIDIOutputElement {
  public:
    PanJoystick(OutputBankConfig<> bankConfig,PinList<2> pins, Channel channel)
      : address(bankConfig,channel),joyX(pins[0]), joyY(pins[1]), channel(channel) {}

   public:
    void begin() final override {
      // Initialize if necessary
    }

    void update() final override {
      bool changed = false;
      changed |= joyX.update();
      changed |= joyY.update();


      if (!changed)
        return;

      constexpr int bias = 512;
      int x = joyX.getValue() - bias; // Position 0 (is actually 512)
      int y = joyY.getValue() - bias; // Position 0 (is actually 512)

      // Math to convert X,Y position to Angle and Distance from the center position
      constexpr float PI_f = M_PI;
      float angleRad          = atan2(y, x);                  // [-π, +π]
      angleRad               += PI_f;                         // [0, +2π]
      float diversity         = hypot(x, y) / bias ; // [0, 1]
      //                        ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
      // Check this formula, it might not do what you expect!
      
      // Map to MIDI CC range
      uint8_t angleMapped     = round(angleRad * 127 / (2 * PI_f)); // [0, 127]
      uint8_t diversityMapped = round(diversity * 127);             // [0, 127]

      // Sending the CC MIDI messages (Both CC must be sent at the same time)
      Control_Surface.sendCC({118, channel}, angleMapped);
      Control_Surface.sendCC({119, channel}, diversityMapped);
    }

  private:
    Bankable::SingleAddress address;
    FilteredAnalog<> joyX, joyY;
    Channel channel;
};
END_CS_NAMESPACE
 
// ::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::: //

USBMIDI_Interface midi;
Bank<2> bank(2);
// Instantiate a Bank selector to control which one of the four Banks is active.
IncrementSelector<2> selector {
  bank, // Bank to manage
  15,    // push button pin
};
 
PanJoystick joystick = {{bank, BankType::ChangeAddress},{28, 27},CHANNEL_1 };


void setup() {
  Control_Surface.begin();
}

void loop() {
  Control_Surface.loop();
}
```

**Comment by tttapa** [comment]
Please provide detailed information of _what_ doesn't work (include compiler errors/warnings, steps to reproduce, descriptions of the behavior you want, and the behavior you observe, etc.).

**Reply by fareign** [reply]
> Please provide detailed information of _what_ doesn't work (include compiler errors/warnings, steps to reproduce, descriptions of the behavior you want, and the behavior you observe, etc.).

no error . cannot change anything (channel or address),I want to change channel or address.

**Comment by fareign** [comment]
`#include "Control_Surface.h"
BEGIN_CS_NAMESPACE

class PanJoystick : public MIDIOutputElement {
  public:
    PanJoystick(OutputBankConfig<> bankConfig,PinList<2> pins,MIDIAddress baseAddress)
      : address(bankConfig,baseAddress),joyX(pins[0]), joyY(pins[1])  {}

   public:
    void begin() final override {
      // Initialize if necessary
    }

    void update() final override {
      bool changed = false;
      changed |= joyX.update();
      changed |= joyY.update();


      if (!changed)
        return;

      constexpr int bias = 512;
      int x = joyX.getValue() - bias; // Position 0 (is actually 512)
      int y = joyY.getValue() - bias; // Position 0 (is actually 512)

      // Math to convert X,Y position to Angle and Distance from the center position
      constexpr float PI_f = M_PI;
      float angleRad          = atan2(y, x);                  // [-π, +π]
      angleRad               += PI_f;                         // [0, +2π]
      float diversity         = hypot(x, y) / bias ; // [0, 1]
      //                        ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
      // Check this formula, it might not do what you expect!
      
      // Map to MIDI CC range
      uint8_t angleMapped     = round(angleRad * 127 / (2 * PI_f)); // [0, 127]
      uint8_t diversityMapped = round(diversity * 1270);             // [0, 127]

      // Sending the CC MIDI messages (Both CC must be sent at the same time)
      //Control_Surface.sendCC({118, channel}, angleMapped);
      Control_Surface.sendCC(address.getActiveAddress(), diversityMapped);
    }

  private:
    Bankable::SingleAddress address;
    FilteredAnalog<> joyX, joyY;

};
END_CS_NAMESPACE
 
// ::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::: //

USBMIDI_Interface midi;
Bank<2> bank(4);
// Instantiate a Bank selector to control which one of the four Banks is active.
IncrementSelector<2> selector {
  bank, // Bank to manage
  15,    // push button pin
};
 
PanJoystick joystick = {{bank, BankType::ChangeChannel},{28, 27}, {118,CHANNEL_16 }};


void setup() {
  Control_Surface.begin();
}

void loop() {
  Control_Surface.loop();
}`

now it can work,but I need declare many address in PanJoystick 
like this:
int ccaddress=118;
if(){
  ccaddress=119;
}elseif {
ccaddress=120;
}
Control_Surface.sendCC({ccaddress, channel}, angleMapped);
but now olny use  Control_Surface.sendCC(address.getActiveAddress(), diversityMapped);  can work.
so I want to split address and channel.

**Comment by fareign** [comment]
now I have the LED trouble.
      switch (address.getSelection()) {
        case 0:

          leds[3] = CRGB::Green;
          FastLED.show();
          break;
        case 1:
          leds[3] = CRGB::Red;
          FastLED.show();
          break;
      }

if I use this ,it works well ,but,fastled use much cpu,so my joystick cant get 0 at center quickly,must slowly comeback to center.

if I add an condition
  if (midiled.getDirty()) {
    FastLED.show();
    midiled.clearDirty();
  }
it notwork, till anorther note key pressed.

if I user a button.update(). it not correct everytime.

      if ( buttonHeld.update() == Button::Falling) {
          FastLED.show(); 
          delay(1000);
          //midiled.clearDirty();
      }

so is there have a good idea?

thanks.
